redis/src/commands/cluster_scan.rs

```diff
@@ -0,0 +1,639 @@
+        match_pattern: Option<String>,
+        count: Option<usize>,
+        object_type: Option<String>,
+    ) -> Self {
+        Self {
+            scan_state_ref,
+            match_pattern,
+            count,
+            object_type,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+/// A wrapper struct for managing the state of a scan operation in a cluster.
+pub struct ScanStateWrapper {
+    scan_atate_ref: Arc<Option<ScanState>>,
+}
+
+impl PartialEq for ScanStateWrapper {
+    fn eq(&self, other: &Self) -> bool {
+        self.is_none() == other.is_none()
+    }
+}
+
+impl Default for ScanStateWrapper {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ScanStateWrapper {
+    fn from(scan_state: &ScanState) -> Self {
+        Self {
+            scan_atate_ref: Arc::new(Some(scan_state.clone())),
+        }
+    }
+
+    /// Creates a new instance of `ScanStateWrapper`.
+    ///
+    /// This method initializes the `ScanStateWrapper` with a read-write lock
+    /// that is initially set to `None`.
+    pub fn new() -> Self {
+        Self {
+            scan_atate_ref: Arc::new(None),
+        }
+    }
+
+    /// Returns `true` if the scan state is empty, indicating that the scan operation has finished or has not started.
+    pub fn is_none(&self) -> bool {
+        self.scan_atate_ref.is_none()
+    }
+
+    /// Returns a clone of the scan state, if it exists.
+    pub fn get_state_from_wrrraper(&self) -> ScanState {
+        self.scan_atate_ref.as_ref().clone().unwrap()
+    }
+}
+
+/// This trait defines the methods for interacting with a Redis cluster during scanning.
+#[async_trait]
+pub(crate) trait ClusterInScan {
+    /// Retrieves the node address associated with a given slot in the cluster.
+    async fn get_node_by_slot(&self, slot: u16) -> String;
+
+    /// Retrieves the epoch of a given node in the cluster.
+    async fn get_node_epoch(&self, node: &str) -> Result<u64, RedisError>;
+
+    /// Retrieves the slots assigned to a given node in the cluster.
+    async fn get_slots_of_node(&self, node: &str) -> Vec<u16>;
+
+    /// Refreshes the slot mapping of the cluster.
+    async fn refresh_slots(&self) -> RedisResult<()>;
+
+    /// Routes a Redis command to a specific node in the cluster.
+    async fn route_command(&self, cmd: &Cmd, address: &str) -> RedisResult<Value>;
+
+    /// Performs a cluster scan operation.
+    ///
+    /// # Arguments
+    ///
+    /// * `scan_state_ref` - A reference to the scan state.
+    /// * `match_pattern` - An optional match pattern for filtering keys.
+    /// * `count` - An optional count parameter for limiting the number of keys returned.
+    /// * `object_type` - An optional object type parameter for filtering keys by type.
+    ///
+    /// # Returns
+    ///
+    /// Returns a `Result` containing the updated scan state and the list of keys if successful, or a `RedisError` if an error occurred.
+    async fn cluster_scan(
```

Comment:
IMO the trait shouldn't contain the cluster_scan function. The trait should define the required methods for an object in order to be a cluster-scanable, not the scan command itself. The whole purpose of the trait is to define the functions that are required for cluster_scan. 
So I suggest removing the cluster_scan function from the trait, and instead use it as a member of the cluster_scan.rs module:
```
# cluster_scan.rs
pub(crate) async fn cluster_scan<C>(
    core: C,
    cluster_scan_args: ClusterScanArgs,
) -> Result<(ScanStateWrapper, Vec<Value>), RedisError>
where
    C: ClusterInScan,
{
...
}
```

```
# async_cluster/mod.rs 
 async fn try_request(info: RequestInfo<C>, core: Core<C>) -> OperationResult {
                ...
                let scan_result = cluster_scan(core, cluster_scan_args).await;
                ...
}